From c484bdd891694532eb74737fee5f0fb2f9650a0a Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Thomas=20L=C3=BCbking?= <thomas.luebking@gmail.com>
Date: Tue, 13 Oct 2015 23:39:50 +0200
Subject: [PATCH] do not write back dated settings from daemon

when the kcm applies, it causes emits of shortcut change signals
this causes a save request in the daemon which then writes its
dated configuration before the kcm asks is to re-read the config

this would usually break changes of only the action or the shortcut

=> we have the kcm mark the updated config file and in the daemon
do not write the settings until it re-read it (where it removes the flag)

this is an IPC race and _not_ related to the previous delaying patch
which would at best mitigate the issue (but the kconfig sync delays too long)

CCBUG: 352067
---
 app/kded.cpp                    |  1 +
 libkhotkeysprivate/settings.cpp | 12 ++++++++++++
 libkhotkeysprivate/settings.h   |  2 ++
 3 files changed, 15 insertions(+)

diff --git a/app/kded.cpp b/app/kded.cpp
index 307d6d4..34eafea 100644
--- a/app/kded.cpp
+++ b/app/kded.cpp
@@ -45,6 +45,7 @@ KHotKeysModule::KHotKeysModule(QObject* parent, const QList<QVariant>&)
     {
     // initialize
     qDebug() << "Installing the delayed initialization callback.";
+    Settings::thisIsTheDaemon = true;
     QMetaObject::invokeMethod( this, "initialize", Qt::QueuedConnection);
     }
 
diff --git a/libkhotkeysprivate/settings.h b/libkhotkeysprivate/settings.h
index b21bdc0..b12622f 100644
--- a/libkhotkeysprivate/settings.h
+++ b/libkhotkeysprivate/settings.h
@@ -181,6 +181,8 @@ public:
      */
     bool isConfigFileValid(KConfigBase const &config, ImportType ask);
 
+    static bool thisIsTheDaemon;
+
 protected:
 
     /**
-- 
2.5.1

diff --git a/libkhotkeysprivate/settings.cpp b/libkhotkeysprivate/settings.cpp
index 43ebe32..6d44173 100644
--- a/libkhotkeysprivate/settings.cpp
+++ b/libkhotkeysprivate/settings.cpp
@@ -36,6 +36,7 @@ namespace KHotKeys
 {
 
 // Settings
+bool Settings::thisIsTheDaemon = false;
 
 Settings::Settings()
     : m_actions( NULL ),
@@ -496,6 +497,8 @@ bool Settings::read_settings(ActionDataGroup *root, KConfigBase const &config, b
             return false;
         }
 
+    if (thisIsTheDaemon)
+        mainGroup.deleteEntry("ConfigUpdated");
     return true;
     }
 
@@ -527,6 +530,15 @@ bool Settings::update()
 void Settings::write()
     {
     KConfig cfg( KHOTKEYS_CONFIG_FILE );
+    if (thisIsTheDaemon && KConfigGroup(&cfg, "Main").readEntry("ConfigUpdated", false))
+        return; // the kcm wrote after we read it and we're trying to write because
+                // we got an update from kglobalaccel - ie. it's pointless and our
+                // info is dated. The kcm should next ask us to read settings
+    if (!thisIsTheDaemon)
+        {
+        KConfigGroup(&cfg, "Main").writeEntry("ConfigUpdated", true);
+        cfg.sync();
+        }
     SettingsWriter writer(this, Retain);
     writer.writeTo(cfg);
     }
